src/plonk/lookup/prover.rs

```diff
@@ -0,0 +1,664 @@
+        //
+        // where there are m input columns and m table columns,
+        // a_j(\omega^i) is the jth input column in this lookup,
+        // a'j(\omega^i) is the permuted input column,
+        // s_j(\omega^i) is the jth table column in this lookup,
+        // s'(\omega^i) is the permuted table column,
+        // and i is the ith row of the column.
+
+        // Compute the evaluations of the lookup product polynomial
+        // over our domain, starting with z[0] = 1
+        let z = iter::once(C::Scalar::one())
+            .chain(lookup_product.into_iter().skip(1))
+            .scan(C::Scalar::one(), |state, cur| {
+                *state *= &cur;
+                Some(*state)
+            })
+            .collect::<Vec<_>>();
+        let z = pk.vk.domain.lagrange_from_vec(z);
+
+        #[cfg(feature = "sanity-checks")]
+        // This test works only with intermediate representations in this method.
+        // It can be used for debugging purposes.
+        {
+            // While in Lagrange basis, check that product is correctly constructed
+            let n = params.n as usize;
+
+            // z'(X) (a'(X) + \beta) (s'(X) + \gamma)
+            // - z'(\omega^{-1} X) (\theta^m a_1(X) + \theta^{m-1} a_2(X) + ... + a_m(X) + \beta) (\theta^m s_1(X) + \theta^{m-1} s_2(X) + ... + s_m(X) + \gamma)
+            for i in 0..n {
+                let prev_idx = (n + i - 1) % n;
+
+                let mut left = z[i];
+                let permuted_input_value = &self.permuted_input_column[i];
+
+                let permuted_table_value = &self.permuted_table_column[i];
+
+                left *= &(*beta + permuted_input_value);
+                left *= &(*gamma + permuted_table_value);
+
+                let mut right = z[prev_idx];
+                let mut input_term = self.unpermuted_input_columns
+                    .iter()
+                    .fold(C::Scalar::zero(), |acc, input| acc * &theta + &input[i]);
+
+                let mut table_term = self.unpermuted_table_columns
+                    .iter()
+                    .fold(C::Scalar::zero(), |acc, table| acc * &theta + &table[i]);
+
+                input_term += &(*beta);
+                table_term += &(*gamma);
+                right *= &(input_term * &table_term);
+
+                assert_eq!(left, right);
+            }
+        }
+
+        let product_blind = Blind(C::Scalar::rand());
+        let product_commitment = params.commit_lagrange(&z, product_blind).to_affine();
+        let z = pk.vk.domain.lagrange_to_coeff(z);
+        let product_coset = pk
+            .vk
+            .domain
+            .coeff_to_extended(z.clone(), Rotation::default());
+        let product_inv_coset = pk.vk.domain.coeff_to_extended(z.clone(), Rotation(-1));
```

Comment:
i've opened https://github.com/zcash/halo2/issues/new. i think `transcript.absorb_point` should remain outside though because the transcript is separate from the `poly` crate.